src/main/kotlin/net/ccbluex/liquidbounce/config/types/ColorValue.kt

```diff
@@ -0,0 +1,119 @@
+ * (at your option) any later version.
+ *
+ * LiquidBounce is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with LiquidBounce. If not, see <https://www.gnu.org/licenses/>.
+ */
+package net.ccbluex.liquidbounce.config.types
+
+import com.google.gson.Gson
+import com.google.gson.JsonObject
+import com.google.gson.annotations.SerializedName
+import net.ccbluex.liquidbounce.config.gson.stategies.Exclude
+import net.ccbluex.liquidbounce.config.gson.stategies.ProtocolExclude
+import net.ccbluex.liquidbounce.event.EventListener
+import net.ccbluex.liquidbounce.event.events.GameRenderEvent
+import net.ccbluex.liquidbounce.event.handler
+import net.ccbluex.liquidbounce.render.engine.Color4b
+import net.ccbluex.liquidbounce.utils.kotlin.EventPriorityConvention
+import org.apache.commons.lang3.StringUtils
+import java.util.*
+import kotlin.math.abs
+import kotlin.math.min
+
+private const val MAX_SPEED = 100
+
+class ColorValue(name: String, defaultValue: Color4b) :
+    Value<Color4b>(name, defaultValue, ValueType.COLOR), CustomDeserializing {
+
+    @SerializedName("rainbowMode")
+    var rainbowMode = RainbowMode.NONE
+
+    @SerializedName("rainbowSpeed")
+    var rainbowSpeed = 20
+
+    @Exclude
+    @ProtocolExclude
+    private var temporaryValue: Color4b? = null
+
+    init {
+        ColorValueManager.values[this] = null
+    }
+
+    override fun get() = temporaryValue ?: super.get()
+
+    override fun deserialize(gson: Gson, jsonObject: JsonObject) {
+        println(jsonObject)
+        rainbowSpeed = jsonObject["rainbowSpeed"]?.asInt?.coerceIn(0..MAX_SPEED) ?: return
+        val modeName = jsonObject["rainbowMode"]?.asString ?: return
+        rainbowMode = RainbowMode.entries.firstOrNull {
+            StringUtils.equalsIgnoreCase(it.choiceName, modeName)
+        } ?: return
+        rainbowMode.init(this)
+    }
+
+    enum class RainbowMode(override val choiceName: String) : NamedChoice {
+
+        NONE("None") {
+
+            override fun update(value: ColorValue, time: Long) {
+                // nothing
+            }
+
+            override fun init(value: ColorValue) {
+                value.temporaryValue = null
+            }
+
+        },
+        CYCLE("Cycle") {
+
+            override fun update(value: ColorValue, time: Long) {
+                val frequency = getFrequency(value)
+                value.temporaryValue = value.inner.hue((time % frequency).toFloat() / frequency)
+            }
+
+        },
+        PULSE("Pulse") {
+
+            override fun update(value: ColorValue, time: Long) {
+                val frequency = getFrequency(value)
+                val oscillation = abs((time % (frequency * 2)) - frequency) / frequency
```

Comment:
maybe add easing. I think something like sine or quadratic easing should look good.

Reply:
I did choose modulo for performance

Reply:
what performance? Calculating a sine takes 40 ns.

Reply:
40ns sounds a bit low unless you have a fast CPU. The JVM can barely optimize this function as it has a new time value each time. Considering it could run on different modules multiple time per frame i think this should be considered (mainly because i had issues with the speed when i implemented that before).

Reply:
One cache miss is as expensive as calculating a sin. If that function runs more than 10 times per frame something is wrong.

Reply:
We need to calculate it for each setting that has it enabled as we can change the speed